### PR TITLE
Update CI, CD

### PR DESCRIPTION
**Pull request by mfep**
- Updated format enforcing script.
- Minor changes to source files to avoid warnings on all compilers tested in CI.
  - `-Werror`/`/WX` is enabled in CI, however a list of warnings had to be enabled, since their resolution was deemed out-of-scope.
- Added `clinfo` as a dependency, that is built with the project.
- Refactored the `CMakeLists.txt` to be able to produce binary Debian packages using `cpack`.
  - CMake minimum version is bumped to 3.16.
- Updated the CI (pre-release) script.
  - Updated Linux CI to use a new container image
    - Most jobs in the matrix are based on Ubuntu 22.04
    - Compilers gcc-11, gcc-13, clang-14, clang-16 and gcc-9 on Ubuntu 20.04
    - Binary targets: x86-64 and x86 on Ubuntu 20.04
    - CMake versions: 3.26 and 3.16 on Ubuntu 20.04
    - CMake generators: Ninja Multi-Config, Unix Makefiles
    - Added test execution
    - Added the generation of the Debian package and testing the consumption of the package, as well as the installed tree
  - Updated Windows CI, which runs on Windows Server 2022
    - MSVC toolsets: v141, v142, v143, clang-cl (this latter with msbuild only)
    - Binary targets: x86-64, x86
    - CMake: system installed, currently 3.27
    - CMake generators: Ninja Multi-Config, Visual Studio 17 2022
    - Added the result checking of the external command invocations in the PowerShell snippets
    - Added test execution
    - Added build and test steps checking the consumption of the installed tree
  - Updated MacOS CI, which runs on MacOS 13
    - Compiler: Apple-Clang 14
    - Binary target: x86-64
    - CMake: system installed, currently 3.27
    - CMake generators: Ninja Multi-Config, XCode
    - Added test execution
    - Added build and test steps checking the consumption of the installed tree
  - Added Android CI, which is build only (virtualization is not available with the default GitHub runners)
    - Binary targets: x86-64, arm64
    - Android API levels: 19, 33
- Updated the release workflow
  - Added `DebSourcePkg.cmake` which is intended to run in CMake script mode. This script generates the `debian/control`, `debian/changelog` and `debian/rules` files which are required to build a Debian source package.
  - The release workflow generates the Debian source package and pushes it to a prescribed Launchpad PPA. The details of the release (e.g. maintainer, signing key) has to be set up as repository variables.
  - `clinfo` is not part of the aforementioned Debian source package, since the Debian package build on Launchpad cannot access the Internet. A potential solution would be forking `clinfo` and referencing it as a submodule, but this was decided to be out-of-scope for now.
  - Detailed instruction regarding the release procedure is added to RELEASE.md.
  
A ToDo item is to revert the referenced submodule commits to the `KhronosGroup` repos, once the respective PRs have been merged.

This PR is considered to be complete, albeit review remarks and/or changes to related PRs might warrant minor updates.

### CODE REVIEW

File: .gitmodules
Comment (bashbaug) [inline .gitmodules:9]: Is this change to the StreamHPC repos intended?  Should this be reverted before merging?

File: docs/RELEASE.md
Comment (bashbaug) [inline docs/RELEASE.md:91]: Could you please add a little more detail what is needed here, and ideally a link to the previous notes?

File: lib/src/Extensions/src/openclext.cpp
Comment (bashbaug) [inline lib/src/Extensions/src/openclext.cpp:71]: Since this is a generated file, if these changes are needed they should be made to the template.  Let me know if you need help with this.

File: lib/src/Utils/File.c
Comment (bashbaug) [inline lib/src/Utils/File.c:341]: What prompted this change?  It should be fine to free a NULL pointer.

File: samples/CMakeLists.txt
Comment (bashbaug) [inline samples/CMakeLists.txt:135]: This means that instead of the samples being installed to `<install_prefix>/install/bin` they will be installed to `<install_prefix>/install/bin/<config>`, so something like `/path/to/sdk/install/bin/Release`.  Was this intended?  I'm not opposed to this change, but it is different than what we had previously.

File: samples/core/binaries/main.c
Comment (bashbaug) [inline samples/core/binaries/main.c:162]: It seems odd to initialize the command queue but not the platform, device, or context.

File: samples/core/blur/blur.cpp
Comment (bashbaug) [inline samples/core/blur/blur.cpp:515]: Nitpick: this helper function seems to be querying whether OpenCL **C** 2.0 is supported, not OpenCL 2.0.

Edit: Is checking for the OpenCL C version the right thing to do here?  Currently the C++ and C samples differ in what they are checking for.

File: samples/core/blur/blur.hpp
Comment (bashbaug) [inline samples/core/blur/blur.hpp:45]: See comment above - I think this is more about OpenCL C 2.0 support, not OpenCL 2.0 support?

File: samples/core/blur/main.c
Comment (bashbaug) [inline samples/core/blur/main.c:1138]: This is incorrect and needs to be fixed: there is no dependency between `cl_khr_subgroup_shuffle_relative` and OpenCL C 2.0.

File: samples/core/blur/main.c
Comment (bashbaug) [inline samples/core/blur/main.c:1150]: This is also incorrect and needs to be fixed: there is no dependency between `cl_khr_subgroup_shuffle` and OpenCL C 2.0.

File: samples/core/blur/main.c
Comment (bashbaug) [inline samples/core/blur/main.c:1195]: See comment above: there is no dependency between `cl_khr_subgroup_shuffle_relative` and OpenCL C 2.0.

File: samples/core/blur/main.c
Comment (bashbaug) [inline samples/core/blur/main.c:1208]: See comment above: there is no dependency between `cl_khr_subgroup_shuffle` and OpenCL C 2.0.

File: samples/core/blur/main.cpp
Comment (bashbaug) [inline samples/core/blur/main.cpp:96]: See comment above: there is no dependency between `cl_khr_subgroup_shuffle_relative` and OpenCL C 2.0.

File: samples/core/blur/main.cpp
Comment (bashbaug) [inline samples/core/blur/main.cpp:106]: See comment above: there is no dependency between `cl_khr_subgroup_shuffle` and OpenCL C 2.0.

File: samples/core/blur/main.cpp
Comment (bashbaug) [inline samples/core/blur/main.cpp:145]: See comment above: there is no dependency between `cl_khr_subgroup_shuffle_relative` and OpenCL C 2.0.

File: samples/core/blur/main.cpp
Comment (bashbaug) [inline samples/core/blur/main.cpp:156]: See comment above: there is no dependency between `cl_khr_subgroup_shuffle` and OpenCL C 2.0.

**Review by bashbaug — CHANGES_REQUESTED** [review]
Note, I did not review all of the CI/CD and CMake changes, but here are some comments about the sample updates themselves.

File: .gitmodules
Comment (MathiasMagnus) [reply .gitmodules:9]: Yes, these should be reverted before merging and are/were only needed while downstream changes were required for clean builds.

File: docs/RELEASE.md
Comment (mfep) [reply docs/RELEASE.md]: Yes, added more elaboration and backlinks to the same document.

File: lib/src/Extensions/src/openclext.cpp
Comment (mfep) [reply lib/src/Extensions/src/openclext.cpp:71]: I've removed it from the `openclext.cpp.mako`, I hope that that is the right file.

For additional context: this change was prompted by a warning fix.

File: samples/core/binaries/main.c
Comment (mfep) [reply samples/core/binaries/main.c:162]: This change was also prompted by a warning fix on Android. Now I looked at it again and realized, that the warning pointed out a valid bug in the resource release logic. That I fixed and this change was reverted.

File: samples/CMakeLists.txt
Comment (Beanavil) [reply samples/CMakeLists.txt:135]: It happened when adding the tests jobs to CI that in some environments the execution of the tests failed when configuring and building for Debug and Release (so first we configured for Debug and Release, then we built for Debug and Release and then we executed), and separating the install of the executables was solving it. Currently I'm not able to reproduce the old buggy behaviour, so probably some of the other changes of this PR triggered the fix for it. We can undo this change then IMHO.

File: lib/src/Utils/File.c
Comment (Beanavil) [reply lib/src/Utils/File.c]: I do not recall what triggered it, but it is true that it should be okay to not check for NULL pointer. I undid this in the [latest commit](https://github.com/KhronosGroup/OpenCL-SDK/pull/88/commits/b3e7aff7e00056508c1051750acb5bc773292f68).

File: samples/CMakeLists.txt
Comment (bashbaug) [reply samples/CMakeLists.txt:135]: If the older behavior still works I think I'd slightly lean towards keeping the same installation directories, but this is something the working group should decide.

File: samples/core/blur/main.cpp
Comment (bashbaug) [inline samples/core/blur/main.cpp:52]: See comment elsewhere regarding checking for OpenCL 2.0 specifically vs. checking for OpenCL 2.0 or newer vs. removing the check completely.

**Review by bashbaug — COMMENTED** [review]
Thanks, many issues are fixed, but a few new ones have cropped up (or I didn't notice them the first time).

I don't think anything is broken with these changes per se, but the checks for OpenCL 2.0 and `cl_khr_subgroups` do mean that these samples cannot run on as many devices as they should.  So, if we want to merge these changes as-is and fix things up afterwards to make progress, I'd be OK with that, though ideally they'd be fixed before merging.

File: samples/core/blur/main.c
Comment (bashbaug) [reply samples/core/blur/main.c:1131]: Heh, I'd actually made this comment previously - at least I'm consistent!

File: samples/CMakeLists.txt
Comment (Beanavil) [reply samples/CMakeLists.txt:135]: Okay, let's see it then in the next WG meeting

File: samples/core/blur/blur.cpp
Comment (Beanavil) [reply samples/core/blur/blur.cpp]: Removed!

File: samples/core/blur/main.c
Comment (Beanavil) [reply samples/core/blur/main.c]: Sorry for overlooking the previous comments! I removed this variable and changed the OpenCL 2.0 version check for one that queries which OpenCL version the device supports, adding the corresponding compiler option (for OpenCL 2.0 and 3.0). This, and propagating correctly the compiler options fixed the weird errors we were getting when not using the OpenCL 2.0 std flag (I think the main issue was the compiler option not being propagated to all the programs run).

File: .gitmodules
Comment (EwanC) [inline .gitmodules:3]: https://github.com/KhronosGroup/OpenCL-ICD-Loader/pull/226 has been merged, can we change this line back?

File: .gitmodules
Comment (EwanC) [inline .gitmodules:9]: https://github.com/KhronosGroup/OpenCL-CLHPP/pull/262 has been merged can we change this line back?

File: .gitmodules
Comment (Beanavil) [reply .gitmodules]: Sure, good catch! Should be updated now

File: .gitmodules
Comment (Beanavil) [reply .gitmodules]: Same as for the comment before, should be updated now

File: cmake/Dependencies/clinfo/ext.h
Comment (EwanC) [inline cmake/Dependencies/clinfo/ext.h:1]: If we're getting the clinfo with FetchContent, why do we need this file which looks like it duplicates `https://github.com/Oblomov/clinfo/blob/master/src/ext.h`? will that be usable when pulled in

File: test/cmake/platformenum.cpp
Comment (EwanC) [inline test/cmake/platformenum.cpp:8]: I understand this change cause semantic differences in C file, it's not to clear what difference this change makes in C++ however. Did something fail without it?

**Review by EwanC — COMMENTED** [review]
I'm not an expert in this SDK repo, so not clicked the big green approve button. But I have looked through this PR to try help unblock getting reviews on it and looks fine to me apart from the few minor comments I've made.

File: cmake/Dependencies/clinfo/ext.h
Comment (Beanavil) [reply cmake/Dependencies/clinfo/ext.h:1]: If I'm not mistaken this is used by the `patch.cmake`, which is passed to the `FetchContent_Declare` of clinfo

File: test/cmake/platformenum.cpp
Comment (Beanavil) [reply test/cmake/platformenum.cpp:8]: This is an old change so I'm not fully sure, but I _think_ it triggered some CI build failures when testing the Windows build with clangcl

File: samples/core/blur/blur.cpp
Comment (bashbaug) [inline samples/core/blur/blur.cpp:523]: I suspect other samples will need to do something similar, so this would be a great SDK utility function.  This code from the OpenCL CTS may be a useful starting point:

https://github.com/KhronosGroup/OpenCL-CTS/blob/cd74e0264391e0424a27f657852fd1e9f10bf42e/test_common/harness/kernelHelpers.cpp#L1645

Strictly speaking, this really ought to be checking for the OpenCL C versions supported by the device, not for the OpenCL version supported by the device, but given the current couplings between the OpenCL version and the OpenCL C versions I _think_ this is OK.

**Review by bashbaug — APPROVED** [review]

**Comment by Beanavil** [comment]
@bashbaug so should we merge?

File: lib/CMakeLists.txt
Comment (ilya-lavrenov) [inline lib/CMakeLists.txt:122]: why was it added here?

The library is already installed above via:
```
 install(
    TARGETS ${UTIL_LIB_TARGET}
    EXPORT OpenCL${UTIL_LIB_NAME}Targets
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
    COMPONENT binary
  )
```

installation w/o ARCHIVE LIBRARY or RUNTIME is bad practice

**Pull request merged**
